documentation: precedence table for the standard library.

Octachron opened the pull request:
Following the disucssion in #1167, this PR proposes to add a copy of the manual's [precedence table](http://caml.inria.fr/pub/docs/manual-ocaml/expr.html#146)  inside the standard library documentation page. This makes possible to link easily the description of each operators to this precedence table, replacing the quite arbitrary ordinal precedence level used before.

To add a bit more context, each operator is also linked to the representative of its precedence class and the representative of the 2-neareast precedence classes.

Unfortunately, ocamldoc does not support table by itself. Thus the table is implemented separately for the latex, html and manpage  backends. Since the texinfo backend seems faulty for the standard library, I did not include a texinfo version of the table.

The last commit is a small fix that document the fact that `#...` are left-associative in the manual .

Octachron committed:
add makefile target for stdlib pdf

Octachron committed:
stdlib documentation: precedence table

Octachron committed:
doc: replace ordinal precedence with links

Octachron committed:
manual: #... operators are left associative

pmetzger commented:
This is a fine thing. It would be nice in theory if the texinfo stuff could be fixed, but I'll admit that even I use it so rarely that I probably wouldn't notice.

gasche commented:
Thanks! Do you have an HTML rendering of the new version?

gasche commented on stdlib/stdlib.mli:
```diff
@@ -1223,6 +1264,104 @@ val at_exit : (unit -> unit) -> unit
    The functions are called in 'last in, first out' order: the
    function most recently added with [at_exit] is called first. *)
 
+(** {1:precedence Appendix: precedence level and associativity of operators }
```
Maybe have an implementation comment here indicating to remember to update the other table (in the manual), and also the other way around?

Octachron replied:
I added the two synchronization comments.

pmetzger commented:
> Thanks! Do you have an HTML rendering of the new version?

Built it, yes. I should mention I noticed a couple of oddities (the hyperlinks for `(&&)` for example come out a different color than the others, and I have no idea why.) Also, in the table itself, you should say "_A_ few other syntactic constructions" rather than "Few other syntactic constructions".  Mostly it seems good. I'm a bit distracted right now so I can't do a tight proofread though.

Octachron committed:
review: typo + synchronisation comment

Octachron commented:
@gasche , [here](https://www.polychoron.fr/ocaml-nonmanual/stdlib_precedence_table/Stdlib.html#precedence) is the new html rendering .

@pmetzger , typo fixed, thanks! Does all `(&&)`s render in a different color?

gasche commented:
It looks weird to have the list of submodules without transition, as if they were part of the appendix. I suppose it was already weird before the appendix was added. Can it be placed before, with a title/section name? Or should there at least be a section for it after the appendix?

Octachron committed:
make precedence appendix an appendix

Octachron commented:
Good point. I moved the appendix to the end of the documentation and added a section for the list of modules of the standard library.

gasche reviewed (approved):
I like the result. Given how easy it is to jump to the table, at least in the HTML version, I'm not completely sure that we really need to recall the levels in the text itself (we could just say `(see the {!reference} for associativity and precedence)`), but I'm content with the current state.

dbuenzli commented:
> I'm not completely sure that we really need to recall the levels in the text itself (we could just say `(see the {!reference} for associativity and precedence)`)

Except for associativity which concerns the operator itself and not its relationship to others, I would tend to agree with @gasche. 

I think precedence relationships among operators is better understood by the table than with the text and the idea of classifying by domain doesn't work that well since for example `@@` and `|>` concern all domains.

Also repeatedly seeing the whole table might help people (at least visual ones) to eventually get it in their brain (personally I'm a lost cause).

Octachron commented:
I agree that the textual description of precedence is not really useful for the html version, I am less sure for the manpage version where it is not possible to follow the link to the table.

pmetzger commented:
> I am less sure for the manpage version where it is not possible to follow the link to the table.

One option is to have a man page specifically for precedence. Some man page readers (including the one I use in emacs) let you click from one man page to another when there are cross references.

Octachron commented:
Indeed, it seems that lifting the table to its own page might be the more portable choice.

gasche commented:
I'm not sure what are valid manpage names, but one that comes to mind is `ocaml-operator-grammar`.

pmetzger commented:
Most Unix systems have an operator(7) man page with C operators and their precedence. An `ocaml-operator` page might be fine.

Octachron committed:
documentation: move operator table to a new page

Octachron commented:
I have split the operator table to its own documentation-only module `ocaml_operators.mld` inside the `stdlib` directory and replaced the `above ... and below ...` indication by a simpler reference

> see {!Ocaml_operators} for more information

gasche commented on manual/manual/refman/expr.etex:
```diff
@@ -126,6 +126,8 @@ The table below shows the relative precedences and associativity of
 operators and non-closed constructions. The constructions with higher
 precedence come first. For infix and prefix symbols, we write
 ``"*"\ldots'' to mean ``any symbol starting with "*"''.
+% Note that this table is duplicated in stdlib/stdlib.mli, these tables should
```
This should be updated, right?

Octachron replied:
Indeed.

gasche commented on stdlib/ocaml_operators.mld:
```diff
@@ -0,0 +1,100 @@
+{%latex:
+%
+% Note: the tables below should be kept in sync with the one in
+% manual/manual/refman/expr.etex .
```
In the LaTeX output (and maybe in the HTML output?), could there be a link to the manual table (and maybe back)?

Octachron replied:
Do you mean adding a link to the absolute path of the table (https://caml.inria.fr/pub/docs/manual-ocaml/expr.html)? Using a relative path sounds difficult since the stdlib documentation is not always generated along the manual.

gasche replied:
Indeed, I had forgotten that the two documents are independent. Let's forget about this idea.

gasche commented:
I don't remember everything; is the PR mergeable in the current state (modulo minor fix discussed above)?

Octachron committed:
documentation: no header for mld files

Octachron commented:
Some check-typo fiddling was missing, but this should be fixed in the last commit.
Otherwise, if the location of the mld file and the reference message are fine, the PR should be mergeable.

gasche commented:
Well then I would suggest that you merge when you are satisfied with the results / typo checks / CI results :-) (I already approved.)

Octachron commented:
I will merge tomorrow to let some time for potential reaction.

Octachron committed:
GPR2117 changes

Octachron committed:
pdf manual:fix refs, compilerlibs/ocaml_operators

Octachron committed:
ocamldoc makefile: clean pdf and texi docs

Octachron committed:
documentation: fix a warning in latex mode

Octachron commented:
I have fixed some missing reference in the pdf manual, and some missing cleanup; I will merge once CIs pass.
